src/lyra/abstract_domains/data_structures/dict_segment_lattice.py

```diff
@@ -0,0 +1,298 @@
+    .. automethod:: DictSegmentLattice._join
+    .. automethod:: DictSegmentLattice._widening
+    """
+
+    # use typing.Generic instead of domains as parameter?
+    def __init__(self, key_domain: Type[Lattice], value_domain: Type[Lattice],
+                 key_d_args: Dict[str, Any] = {}, value_d_args: Dict[str, Any] = {},
+                 segments: Set[Tuple[Lattice, Lattice]] = None):
+        """
+        :param key_domain: domain for abstraction of dictionary keys,
+            ranges over the scalar variables and the special key variable v_k
+        :param value_domain: domain for abstraction of dictionary values,
+            ranges over the scalar variables and the special value variable v_v
+        :param key_d_args: arguments for the key_domain
+        :param value_d_args: arguments for the value_domain
+        :param segments: initial set of abstract segments; if None, a top segment is created
+        """
+        # TODO: segments internally as List?
+        super().__init__()
+        self._k_domain = key_domain
+        self._k_d_args = key_d_args
+        self._v_domain = value_domain
+        self._v_d_args = value_d_args
+
+        if segments is None:    # default element
+            self._segments = {(key_domain(**key_d_args).top(), value_domain(**value_d_args).top())}
+        else:
+            # for s in segments:      # TODO: remove?
+            #     if len(s) != 2:
+            #         raise TypeError(f"Segment {s} needs to have exactly two elements,
+            #                         but has {len(s)}")
+            #     if not isinstance(s[0], key_domain):
+            #         raise TypeError(f"The key type of the segment {s}
+            #                         does not match the provided key domain {key_domain}")
+            #     if not isinstance(s[1], value_domain):
+            #         raise TypeError(f"The value type of the segment {s}
+            #                         does not match the provided value domain {value_domain}")
+            self._segments = segments
+
+    @property
+    def k_domain(self) -> Type[Lattice]:
+        """Domain for the abstract keys."""
+        return self._k_domain
+
+    @property
+    def v_domain(self) -> Type[Lattice]:
+        """Domain for the abstract values."""
+        return self._v_domain
+
+    @property
+    def k_d_args(self):
+        """Arguments for the key domain given in constructor"""
+        return self._k_d_args
+
+    @property
+    def v_d_args(self):
+        """Arguments for the value domain given in constructor"""
+        return self._v_d_args
+
+    @property
+    def segments(self) -> Set[Tuple[Lattice, Lattice]]:
+        """Set of all abstract segments."""
+        return self._segments
+
+    def sorted_segments(self) -> List[Tuple[Lattice, Lattice]]:
+        """List of all abstract segments ordered by their keys"""
+        return sorted(self.segments, key=lambda t: t[0])
```

Comment:
This has no reason to exist as a separate method (where it is currently used you are breaking encapsulation and that should not happen).